### PR TITLE
O-RAN V3 Rest Api: Event Subscription

### PR DESCRIPTION
**Pull request by jzding**

### CODE REVIEW

**Review by josephdrichard — COMMENTED** [review]
/lgtm

File: pkg/restclient/client.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+// PostEvent post an event to the give url and check for error
+func (r *Rest) PostCloudEvent(url *types.URI, e ce.Event) (status int, err error) {
+	b, err := json.Marshal(e)
+	if err != nil {
+		log.Errorf("error marshalling event %v", e)
+		return status, err
+	}
+
+	if status = r.Post(url, b); status == http.StatusBadRequest {
```
Comment (aneeshkp) [inline pkg/restclient/client.go:58]: posting of cloud events should be wrapped with O-Ran cloudevents API,. Why rest API is posting cloudEvents object ?

Comment (jzding) [reply pkg/restclient/client.go:58]: O-RAN spec uses Cloud Events as Event Data Model.

> 7.2 Status Notifications Data Model
> This clause specifies the event Status Notification data model supported by the API. The current model supports JSON encoding of the CloudEvents.io specification [15] for the event payload.

File: v2/routes.go [hunk 2]
```diff
@@ -52,50 +50,125 @@ import (
...
+		return
 	}
-	// check sub.EndpointURI by get
-	sub.SetID(uuid.New().String())
-	_ = sub.SetURILocation(fmt.Sprintf("http://localhost:%d%s%s/%s", s.port, s.apiPath, "subscriptions", sub.ID)) //nolint:errcheck
+	if subExists, ok := s.pubSubAPI.HasSubscription(sub.GetResource()); ok {
+		respondWithStatusCode(w, http.StatusConflict,
+			fmt.Sprintf("subscription (id: %s) with same resource already exists, skipping creation",
+				subExists.GetID()))
+		localmetrics.UpdateSubscriptionCount(localmetrics.FAILCREATE, 1)
```
Comment (aneeshkp) [inline v2/routes.go:74]: remove increase

File: v2/routes.go [hunk 3]
```diff
@@ -52,50 +50,125 @@ import (
...
+		return
+	}
+
+	id := uuid.New().String()
+	sub.SetID(id)
+	sub.SetVersion(API_VERSION)
+	sub.SetURILocation(fmt.Sprintf("http://localhost:%d%s%s/%s", s.port, s.apiPath, "subscriptions", sub.ID)) //nolint:errcheck
 
 	newSub, err := s.pubSubAPI.CreateSubscription(sub)
```
Comment (aneeshkp) [inline v2/routes.go:83]: don't need local pubsub anymore, since we are using configmap

Comment (aneeshkp) [reply v2/routes.go:83]: delete this

Comment (jzding) [reply v2/routes.go:83]: added note, will do in the next story

**Pull request merged**